nixos/lib/testing/matrix.nix

```diff
@@ -0,0 +1,248 @@
+{ config, lib, options, extendModules, moduleType, ... }:
+let
+  inherit (lib) mkOption optionalAttrs types mdDoc;
+
+  extendModule = module: extendModules { modules = [ module ]; };
+
+  extend = module: config.callTest ((extendModule module).config.result);
+
+  decisionModule = decision@{ name, ... }: {
+    options = {
+      after = mkOption {
+        description = mdDoc ''
+          > You've already made the choice, now you have to understand it.
+
+          Some choices depend on other choices.
+
+          For example, we wouldn't be able to decide about the cookie before
+          we take the pill. Deciding the cookie is not possible without
+          crucial values set by the pill. To avoid an error like `The option
+          'pill' is used but not defined`, we force the ordering.
+
+          ```nix
+            matrix.cookie = {
+              after = ["pill"];
+            };
+          ```
+        '';
+        type = types.listOf types.str;
+        default = [ ];
+      };
+      enable = mkOption {
+        type = types.bool;
+        default = true;
+        description = mdDoc ''
+          > But if you already know, how can I make a choice?
+
+          > Because you didn't come here to make the choice, you've already made it. You're here to try to understand why you made it.
+
+          Some choices don't make sense, because of earlier choices you've made.
+
+          With this option, you can make Nix understand that, pruning nonsensical
+          decisions from the returned tree of tests, so that we don't need to
+          compute more than necessary.
+
+          ```nix
+          matrix.cookie = {
+            after = ["pill"];
+            enable = config.pill == "red";
+            # ...
+          ```
+
+          If we chose the blue pill, we would never have to make a choice about
+          the cookie, so we skip this choice.
+        '';
+      };
+      choice = mkOption {
+        type = types.lazyAttrsOf (types.submodule (choiceModule decision.name extendModules));
+        description = ''
+          A attribute set where each attribute name corresponds to a choice; a single outcome of the decision.
+
+          When you define `matrix.foo.choice.bar`, the (new) module argument `foo` will be set to `bar` in a reinstantiation of the test configuration that produces the final test.
```

Comment:
I don't unterstand this. How does `foo` become a (new - what does it mean to be new here?) argument? Can you show an example?